File: metisfl/controller/core/model_manager.cc
#include "metisfl/controller/core/model_manager.h"

namespace metisfl::controller {

// Constructor
ModelManager::ModelManager(LearnerManager *learner_manager, Selector *selector,
                           const GlobalTrainParams &global_train_params,
                           const ModelStoreParams &model_store_params)
    : model_(), global_train_params_(global_train_params) {
  learner_manager_ = learner_manager;
  selector_ = selector;
  model_store_ = CreateModelStore(model_store_params);
  aggregator_ = CreateAggregator(global_train_params);
}

// Public methods

absl::Status ModelManager::SetInitialModel(const Model &model) {
  if (is_initialized_)
    return absl::FailedPreconditionError("Model is already initialized.");

  model_ = model;
  is_initialized_ = true;
  return absl::OkStatus();
}

void ModelManager::InsertModel(std::string &learner_id, const Model &model) {
  model_store_->InsertModel(std::vector<std::pair<std::string, Model>>{
      std::pair<std::string, Model>(learner_id, model)});
}

void ModelManager::EraseModels(std::vector<std::string> &learner_id) {
  model_store_->EraseModels(learner_id);
}

void ModelManager::UpdateModel(std::vector<std::string> &to_schedule,
                               std::vector<std::string> &learner_ids) {
  std::vector<std::string> selected_ids =
      selector_->Select(to_schedule, learner_ids);

  auto scaling_factors = ComputeScalingFactors(selected_ids);

  auto update_id = InitializeMetadata();
  int stride_length = GetStrideLength(learner_ids.size());

  std::chrono::time_point<std::chrono::system_clock> start_time_aggregation =
      std::chrono::high_resolution_clock::now();

  for (std::string learner_id : learner_ids) {
    auto lineage_length = GetLineageLength(learner_id);
    std::vector<std::pair<std::string, int>> to_select_block;

    to_select_block.emplace_back(learner_id, lineage_length);
    int block_size = to_select_block.size();

    if (block_size == stride_length || learner_id == learner_ids.back()) {
      auto selected_models = SelectModels(update_id, to_select_block);

      auto to_aggregate_block =
          GetAggregationPairs(selected_models, scaling_factors);

      Aggregate(update_id, to_aggregate_block);

      RecordBlockSize(update_id, block_size);

      model_store_->ResetState();
    }
  }

  RecordAggregationTime(update_id, start_time_aggregation);
  aggregator_->Reset();
}

void ModelManager::Shutdown() { model_store_->Shutdown(); }

absl::flat_hash_map<std::string, double> ModelManager::ComputeScalingFactors(
    const std::vector<std::string> &selected_ids) const {
  auto scaling_factor = global_train_params_.scaling_factor;

  if (scaling_factor == "NumCompletedBatches") {
    auto num_completed_batches =
        learner_manager_->GetNumCompletedBatches(selected_ids);
    return Scaling::GetBatchesScalingFactors(num_completed_batches);
  } else if (scaling_factor == "NumParticipants") {
    return Scaling::GetParticipantsScalingFactors(selected_ids);
  } else if (scaling_factor == "NumTrainingExamples") {
    auto num_training_examples =
        learner_manager_->GetNumTrainingExamples(selected_ids);
    return Scaling::GetDatasetScalingFactors(num_training_examples);
  } else {
    LOG(FATAL) << "Unsupported scaling factor.";
  }
}

std::string ModelManager::InitializeMetadata() {
  auto update_id = metisfl::controller::GenerateRadnomId();
  metadata_[update_id] = ModelMetadata();
  return update_id;
}

// Private methods
int ModelManager::GetStrideLength(int num_learners) const {
  uint32_t stride_length = num_learners;
  if (global_train_params_.aggregation_rule == "FedStride") {
    auto fed_stride_length = global_train_params_.stride_length;
    if (fed_stride_length > 0) {
      stride_length = fed_stride_length;
    }
  }
  return stride_length;
}

int ModelManager::GetLineageLength(std::string &learner_id) const {
  const auto lineage_length = model_store_->GetLearnerLineageLength(learner_id);

  auto required_lineage_length = aggregator_->RequiredLearnerLineageLength();

  return std::min(lineage_length, required_lineage_length);
}

std::map<std::string, std::vector<const Model *>> ModelManager::SelectModels(
    std::string &update_id,
    std::vector<std::pair<std::string, int>> &to_select_block) {
  auto start_time_selection = std::chrono::high_resolution_clock::now();

  std::map<std::string, std::vector<const Model *>> selected_models =
      model_store_->SelectModels(to_select_block);

  auto end_time_selection = std::chrono::high_resolution_clock::now();

  std::chrono::duration<double, std::milli> elapsed_time_selection =
      end_time_selection - start_time_selection;

  metadata_[update_id].set_selection_duration_ms(
      elapsed_time_selection.count());

  return selected_models;
}

std::vector<std::vector<std::pair<const Model *, double>>>
ModelManager::GetAggregationPairs(
    std::map<std::string, std::vector<const Model *>> &selected_models,
    absl::flat_hash_map<std::string, double> &scaling_factors) const {
  std::vector<std::vector<std::pair<const Model *, double>>> to_aggregate_block;

  std::vector<std::pair<const Model *, double>> to_aggregate_learner_models_tmp;

  for (auto const &[selected_learner_id, selected_learner_models] :
       selected_models) {
    auto scaling_factor = scaling_factors[selected_learner_id];

    for (auto it : selected_learner_models) {
      to_aggregate_learner_models_tmp.emplace_back(it, scaling_factor);
    }

    to_aggregate_block.push_back(to_aggregate_learner_models_tmp);
    to_aggregate_learner_models_tmp.clear();
  }

  return to_aggregate_block;
}

void ModelManager::Aggregate(
    std::string &update_id,
    std::vector<std::vector<std::pair<const Model *, double>>>
        &to_aggregate_block) {
  auto start_time_block_aggregation = std::chrono::high_resolution_clock::now();

  model_ = aggregator_->Aggregate(to_aggregate_block);

  auto end_time_block_aggregation = std::chrono::high_resolution_clock::now();

  std::chrono::duration<double, std::milli> elapsed_time_block_aggregation =
      end_time_block_aggregation - start_time_block_aggregation;

  *metadata_[update_id].mutable_aggregation_block_duration_ms()->Add() =
      elapsed_time_block_aggregation.count();
}

void ModelManager::RecordBlockSize(std::string &update_id, int block_size) {
  *metadata_[update_id].mutable_aggregation_block_size()->Add() = block_size;
  long block_memory = GetTotalMemory();
  *metadata_[update_id].mutable_aggregation_block_memory_kb()->Add() =
      (double)block_memory;
}

void ModelManager::RecordAggregationTime(
    std::string &update_id,
    std::chrono::time_point<std::chrono::system_clock> &start) {
  auto end = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double, std::milli> elapsed = end - start;
  metadata_[update_id].set_aggregation_duration_ms(elapsed.count());
}

};  // namespace metisfl::controller

File: metisfl/server/core/model_manager.py

import time
from typing import Dict, List, Tuple
from metisfl.common.types import GlobalTrainConfig
from metisfl.proto import controller_pb2, model_pb2
from metisfl.common.utils import random_id_generator
from metisfl.server.aggregation import Aggregator
from metisfl.server.core import LearnerManager
from metisfl.server.selection.scheduled_cardinality import ScheduledCardinality
from metisfl.server.store import ModelStore
from metisfl.server.scaling import batches_scaling, participants_scaling, dataset_scaling


class ModelManager:

    aggregator: Aggregator = None

    model: model_pb2.Model = None
    model_store: ModelStore = None
    global_train_config: GlobalTrainConfig = None
    metadata: controller_pb2.ModelMetadata = {}
    selector: ScheduledCardinality = None
    learner_manager: LearnerManager = None
    is_initialized: bool = False

    def __init__(
        self,
        global_train_config: GlobalTrainConfig,
        learner_manager: LearnerManager,
        model_store: ModelStore,
    ) -> None:
        """Initializes the model manager.

        Parameters
        ----------
        learner_manager : LearnerManager
            The learner manager to be used.
        global_train_config : GlobalTrainConfig
            The global training configuration.
        model_store_config : ModelStoreConfig
            The model store configuration.
        """
        self.learner_manager = learner_manager
        self.global_train_config = global_train_config
        self.model_store = model_store

    def set_initial_model(self, model: model_pb2.Model) -> None:
        """Sets the initial model for the controller.

        Parameters
        ----------
        model : model_pb2.Model
            The initial model.
        """

        if self.is_initialized:
            raise Exception("Model already initialized")

        self.model = model
        self.is_initialized = True

    def insert_model(self, learner_id: str, model: model_pb2.Model) -> None:
        """Inserts a model into the model manager.

        Parameters
        ----------
        learner_id : str
            The learner id.
        model : model_pb2.Model
            The model to be inserted.
        """
        self.model_store.insert(
            [(learner_id, model)],
        )

    def update_model(
        self,
        to_schedule: List[str],
        learner_ids: List[str],
    ) -> None:
        """Updates the model performing an aggregation step.

        Parameters
        ----------
        to_schedule : List[str]
            The learners to be scheduled.
        learner_ids : List[str]
            The learners ids.
        """
        selected_ids = self.selector.select(to_schedule, learner_ids)
        scaling_factors = self.compute_scaling_factor(learner_ids)
        stride_length = self.get_stride_length(len(learner_ids))

        update_id = self.init_metadata()
        aggregation_start_time = time.time()

        # FIXME: continue this

    def erase_models(self, learner_ids: List[str]) -> None:
        """Erases the models of the learners.

        Parameters
        ----------
        learner_ids : List[str]
            The learners ids.
        """
        self.model_store.erase(learner_ids)

    def get_model(self) -> model_pb2.Model:
        """Gets the model.

        Returns
        -------
        model_pb2.Model
            The model.
        """
        return self.model

    def init_metadata(self) -> str:
        """Initializes the metadata."""

        update_id = str(random_id_generator())
        self.metadata[update_id] = controller_pb2.ModelMetadata()

        return

    def get_stride_length(self, num_learners: int) -> int:
        """Returns the stride length.

        Parameters
        ----------
        num_learners : int
            The number of learners.

        Returns
        -------
        int
            The stride length.
        """

        stride_length = num_learners

        if self.global_train_config.aggregation_rule == "FedStride":
            fed_stride_length = self.global_train_config.stride_length
            if fed_stride_length > 0:
                stride_length = fed_stride_length

        return stride_length

    def get_lineage_length(self, learner_id: str) -> int:
        """Returns the lineage length for the given learner id.

        Parameters
        ----------
        learner_id : str
            The learner id.

        Returns
        -------
        int
            The lineage length for the given learner id.
        """
        pass

    def compute_scaling_factor(
        self,
        learner_ids: List[str],
    ) -> Dict[str, float]:
        """Computes the scaling factor for the given learners.

        Parameters
        ----------
        learner_ids : List[str]
            The learner ids.

        Returns
        -------
        Dict[str, float]
            The scaling factor for each learner.
        """

        scaling_factor = self.global_train_config.scaling_factor

        if scaling_factor == "NumCompletedBatches":
            num_completed_batches = self.learner_manager.get_num_completed_batches(
                learner_ids
            )
            return batches_scaling(num_completed_batches)
        elif scaling_factor == "NumTrainingExamples":
            num_training_examples = self.learner_manager.get_num_training_examples(
                learner_ids
            )
            return dataset_scaling(num_training_examples)
        elif scaling_factor == "NumParticipants":
            return participants_scaling(learner_ids)
        else:
            raise Exception("Invalid scaling factor")

    def get_aggregation_pairs(
        self,
        selected_models: Dict[str, List[model_pb2.Model]],
        scaling_factors: Dict[str, float],
    ) -> List[List[Tuple[model_pb2.Model, float]]]:
        """Returns the aggregation pairs. """

    def aggregate(
        self,
        update_id: str,
        to_aggregate_block: List[List[Tuple[model_pb2.Model, float]]],
    ) -> None:
        """Aggregates the models.

        Parameters
        ----------
        update_id : str
            The update id.
        to_aggregate_block : List[List[Tuple[model_pb2.Model, float]]]
            The models to be aggregated.
        """
        pass

    def record_block_size(self, update_id: str, block_size: int) -> None:
        """Records the block size.

        Parameters
        ----------
        update_id : str
            The update id.
        block_size : int
            The block size.
        """
        pass

    def record_aggregation_time(self, update_id: str, start_time: float) -> None:
        """Records the aggregation time.

        Parameters
        ----------
        update_id : str
            The update id.
        start_time : float
            The start time.
        """
        pass

    def record_model_size(self, update_id: str) -> None:
        """Records the model size.

        Parameters
        ----------
        update_id : str
            The update id.
        """
        pass
